### ISSUE TITLE
Prevent file fragmentation on Windows by pre-allocating files

### ISSUE THREAD

**Issue by mehrdadn** [ncw/rclone#2469]
Hi! Thank you for the awesome program! I just wanted to make a feature request.  

I've noticed that, on Windows, files copied locally can become heavily fragmented (300+ fragments for maybe 100 MB), no matter how much contiguous free space there is (even if it's over 1TiB). This can needlessly yet severely adversely affect performance on hard disks.

There is an easy solution: after creation, simply call [`NtSetInformationFile(..., FileAllocationInformation, ...)`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntifs/nf-ntifs-ntsetinformationfile) in NTDLL.DLL to pre-allocate space for the file. Note that this size must be a multiple of the cluster size; the cluster size can be obtained via [`NtQueryVolumeInformationFile(..., FileFsSizeInformation, ...)`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntifs/nf-ntifs-ntqueryvolumeinformationfile). Also note that the allocation size does not affect the file size; it must simply be at least as large as the file size. I believe these changes would go in the `Update` procedure of the `backend/local/local.go` file.

Thank you!

**Comment by ncw** [comment]
That seems like a good idea.

Rclone doesn't always know the size of the incoming file but where it does it could call that easily enough.

Is this something you could work on? I'm happy to give help where required!

**Comment by ncw** [comment]
@mehrdadn wrote:

> Yeah I pieced together a working implementation that I can share here actually.

Great!

> I just haven't tried to handle edge cases (like non-Windows code, errors when getting the size, or others I haven't thought of).

OK.  Since this is a performance enhancement, rclone should emit a debug or warning and carry on without the feature if anything errors.

> Usually when I try to contribute code though people have requirements/hoops I'm not inclined to jump through, like putting my name or email in the repo... do you have any particular requirements? (License-wise I don't mind public-domaining the code so that part shouldn't be a problem.)

What the rclone project does is put whatever you put in your git commit log (which is public) into the [authors.md](https://github.com/ncw/rclone/blob/master/docs/content/authors.md) file.  As you can see there it doesn't have to be your real name and people often use a github email address.

So if you are happy with that then send a pull request and I'll review the code.  If not then what I suggest we do is that you send me a `git diff` via email to nick@craig-wood.com (put a link to this issue in the body please) and I'll tidy it up and merge it under my own credentials with the understanding that I have your permission.  I'll credit @mehrdadn in the commit message.

Rclone uses the [MIT licence](https://github.com/ncw/rclone/blob/master/COPYING) which is only a very small step away from public domain.

**Comment by mehrdadn** [comment]
Well it seem like in either of those cases I'd still have to inject some personal info (like my username, which BTW is changeable and could point to someone else in the future) into the repo, which is what I don't want to do (even with already-public information). But if you're willing to avoid including my info, you're welcome to use my code here. Since I didn't know Go, I got some trivial bits of it from [here](/Microsoft/hcsshim/blob/0a4192c603327c6ba07bd5dfa3fd1389ee418c56/internal/safefile/zsyscall_windows.go#L56) and [here](/hillu/go-ntdll/blob/776f58ba99f0d3f13111d2e538d23725de4f82f1/file_generated.go#L78), which seemed to have been machine-generated in projects under the MIT and BSD-2 licenses. Everything else is what I wrote, which I release here as public domain so you can use it all freely. =) I don't want any credit so if you don't include anything (or even claim credit yourself) that's perfectly fine by me. But if you absolutely have to put something, you can just put the URL of this page or comment. Hope this helps!

	diff --git a/backend/local/local.go b/backend/local/local.go
	index 1be99a14..aa886ba5 100644
	--- a/backend/local/local.go
	+++ b/backend/local/local.go
	@@ -12,8 +12,10 @@ import (
		"runtime"
		"strings"
		"sync"
	+	"syscall"
		"time"
		"unicode/utf8"
	+	"unsafe"
	 
		"github.com/ncw/rclone/fs"
		"github.com/ncw/rclone/fs/config/configmap"
	@@ -21,6 +23,7 @@ import (
		"github.com/ncw/rclone/fs/hash"
		"github.com/ncw/rclone/lib/readers"
		"github.com/pkg/errors"
	+	"golang.org/x/sys/windows"
	 )
	 
	 // Constants
	@@ -788,6 +791,40 @@ func (o *Object) mkdirAll() error {
		return os.MkdirAll(dir, 0777)
	 }
	 
	+type FILE_ALLOCATION_INFORMATION struct {
	+	AllocationSize uint64
	+}
	+type FILE_FS_SIZE_INFORMATION struct {
	+	TotalAllocationUnits uint64 
	+	AvailableAllocationUnits uint64
	+	SectorsPerAllocationUnit uint32
	+	BytesPerSector uint32
	+}
	+var (
	+	ntdll = windows.NewLazySystemDLL("ntdll.dll")
	+	NtQueryInformationFile = ntdll.NewProc("NtQueryInformationFile")
	+	NtQueryVolumeInformationFile = ntdll.NewProc("NtQueryVolumeInformationFile")
	+	NtSetInformationFile = ntdll.NewProc("NtSetInformationFile")
	+)
	+type ioStatusBlock struct {
	+	Status, Information uintptr
	+}
	+func ntQueryInformationFile(handle uintptr, iosb *ioStatusBlock, information uintptr, length uint32, class uint32) (status uint32) {
	+	r0, _, _ := syscall.Syscall6(NtQueryInformationFile.Addr(), 5, uintptr(handle), uintptr(unsafe.Pointer(iosb)), uintptr(information), uintptr(length), uintptr(class), 0)
	+	status = uint32(r0)
	+	return
	+}
	+func ntQueryVolumeInformationFile(handle uintptr, iosb *ioStatusBlock, information uintptr, length uint32, class uint32) (status uint32) {
	+	r0, _, _ := syscall.Syscall6(NtQueryVolumeInformationFile.Addr(), 5, uintptr(handle), uintptr(unsafe.Pointer(iosb)), uintptr(information), uintptr(length), uintptr(class), 0)
	+	status = uint32(r0)
	+	return
	+}
	+func ntSetInformationFile(handle uintptr, iosb *ioStatusBlock, information uintptr, length uint32, class uint32) (status uint32) {
	+	r0, _, _ := syscall.Syscall6(NtSetInformationFile.Addr(), 5, uintptr(handle), uintptr(unsafe.Pointer(iosb)), uintptr(information), uintptr(length), uintptr(class), 0)
	+	status = uint32(r0)
	+	return
	+}
	+
	 // Update the object from in with modTime and size
	 func (o *Object) Update(in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) error {
		hashes := hash.Supported
	@@ -807,7 +844,17 @@ func (o *Object) Update(in io.Reader, src fs.ObjectInfo, options ...fs.OpenOptio
		if err != nil {
			return err
		}
	-
	+	var iosb ioStatusBlock
	+	var fsSizeInfo FILE_FS_SIZE_INFORMATION
	+	if ntQueryVolumeInformationFile(out.Fd(), &iosb, uintptr(unsafe.Pointer(&fsSizeInfo)), uint32(unsafe.Sizeof(fsSizeInfo)), 3) == 0 {
	+		var allocInfo FILE_ALLOCATION_INFORMATION
	+		size := src.Size()
	+		if size > 0 {
	+			clusterSize := uint64(fsSizeInfo.BytesPerSector) * uint64(fsSizeInfo.SectorsPerAllocationUnit)
	+			allocInfo.AllocationSize = (1 + uint64(uint64(src.Size() - 1) / clusterSize)) * clusterSize
	+			ntSetInformationFile(out.Fd(), &iosb, uintptr(unsafe.Pointer(&allocInfo)), uint32(unsafe.Sizeof(allocInfo)), 19)
	+		}
	+	}
		// Calculate the hash of the object we are reading as we go along
		hash, err := hash.NewMultiHasherTypes(hashes)
		if err != nil {

**Comment by ncw** [comment]
@mehrdadn thanks for that :-)  I'll have a go at integrating it.  I'll put an issue marker on the commit #2469 but other than that I'll leave you anonymous.

**Comment by mehrdadn** [comment]
Awesome, thanks! :)

**Comment by ncw** [comment]
I've fixed the error handling, made it cross platform and stuck it in a branch

Here is the commit: 99603ce1f4c2d820017c11c3e59ce848d85e4eec

@mehrdadn would you mind giving the commit a review and running some tests to see if it OK?

Thank you :-)

There is a binary too here:

https://beta.rclone.org/branch/v1.42-113-g99603ce1-fix-2469-win-frag/ (uploaded in 15-30 mins)

**Comment by mehrdadn** [comment]
@ncw There's just one potential pitfall: if you ever use `preAllocate` in the future and forget not to pass it `size == 0`, it will fail. You might want to change `(1 + uint64(size-1)/clusterSize) * clusterSize`  to be conditional on `size > 0`. Other than that it looks great. :)

**Comment by ncw** [comment]
@mehrdadn good idea - I've moved the check into the preallocate function instead.

I've also been investigating the Linux equivalent [fallocate](http://man7.org/linux/man-pages/man2/fallocate.2.html) - I don't know whether that is worth implementing though as I think linux filesystems are very good at keeping fragmentation under control.

**Comment by ncw** [comment]
@mehrdadn I agree with you about the POSIX `fallocate`.  Linux seems to have some linux only extensions which appear to do the right thing, ie allocate blocks for the file, which are zeroed when read but it doesn't take O(n) time to do so.

The flag in question is called `FALLOC_FL_KEEP_SIZE`

> If the FALLOC_FL_KEEP_SIZE flag is specified in mode, the behavior of
> the call is similar, but the file size will not be changed even if
> offset+len is greater than the file size.  Preallocating zeroed
> blocks beyond the end of the file in this manner is useful for
> optimizing append workloads.

What do you think of that?

**Comment by mehrdadn** [comment]
Any news on this? Could I suggest that you just merge the Windows version and do the other platforms in a separate changeset?

**Comment by ncw** [comment]
Thanks for giving me a nudge on this, it had fallen off my radar!

I'll merge the code so far in a bit.

**Comment by ncw** [comment]
I've merged this now  - it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins.

It has linux support too.

Thank you very much for your help.

**Comment by ferferga** [comment]
Is there any flag to disable this behaviour? @ncw @mehrdadn

**Comment by ncw** [comment]
> Is there any flag to disable this behaviour? @ncw @mehrdadn

No but there could be. What problem is it causing you? Did you try [the latest beta](https://beta.rclone.org) which has some fixes in this area?

**Comment by ferferga** [comment]
@ncw Nope, I haven't tried it. Can you point me to the specific commits/PRs where stuff with it has been done?

The problem I mainly have with it is that, if I download a huge file (like 50 GB) and I'm doing other operations in the HDD disk (of course SSDs are not that affected), rclone takes like double the time it takes for my bandwidth to download the file just to preallocate it.

I know that, if I'm doing other operations in my HDD, stuff is going to be slow in my drive. However, many of the time I get this annoyance the drive is still capable of getting 10 MB/s, which is exactly what my connection provides. Of course, if the drive is still more sluggish than that, rclone will throttle, but I want that: throttle it instead. Because otherwise, what I'm doing is just writing twice the same amount of data.

This is even worse when you have more than 5 transfers of a couple gigabytes each. Rclone will fire 5 threads writing to  the disk, so it will be all random reads, which really kills performance. Today, I took like an hour to download 20 GB, something that shouldn't take more than 20 minutes.

There are also two problems with the implementation, which I even thought they were bugs:

* **No visual indication**: If you're using ``--progress``, you will see that the file is downloading at 16, kb/s and it's completely stuck. No visual state like "preallocating" is displayed.

* **Doesn't shutdown gracefully**: While preallocating is being done, doing CTRL+C completely breaks it. It wants first to finish and, once it's finished, it deletes the allocated area and closes itself.

This happened many times to me with big files (around 400 GB in size). When that happens, rclone isn't even killable by Task Manager, and the whole Explorer and many apps became unresponsive. This isn't an rclone issue per-se, it's cause basically because Windows isn't that good with programs locking drives and files (it happens a lot when you use mount, open a file with a program and there is lag between the remote and your computer. The whole app is unresponsive when that happens and it can't be killed by Task Manager because it's locked to a file, and somewhat Windows thinks that's unsafe to kill).

I think the current implementation should be done, of course, deactivable and dynamic. Doesn't make so much sense imo to allocate everything at first: if, for example, I'm downloading at an average of 12 MB/s, rclone can be allocating ahead at an 1.5X the average speed (so 18 MB/s). That could make the situation better for large files and, if the storage medium is slow, have some "spare space" ahead to even make the transfer faster. I'm not an expert by any means, so maybe my suggestion has more inconveniences than advantages.

Thank you very much, as always, for your help! @ncw

**Comment by ncw** [comment]
> The problem I mainly have with it is that, if I download a huge file (like 50 GB) and I'm doing other operations in the HDD disk (of course SSDs are not that affected), rclone takes like double the time it takes for my bandwidth to download the file just to preallocate it.

This is exactly what got fixed in the beta. It isn't the pre-allocation it is the writing to not the start of the file that is the problem. The beta sets the file so it can be sparse on Windows

a492c0fb0e2c84bbdb8d7612ba3fa1bfce4dab8f

**Comment by mehrdadn** [comment]
Interesting. Does making it sparse not cause the same fragmentation we were trying to avoid?

**Comment by ferferga** [comment]
@ncw This is exactly how all other programs usually work. Thank you very much for this, it's great news!

**Comment by ncw** [comment]
@mehrdadn wrote:
> Interesting. Does making it sparse not cause the same fragmentation we were trying to avoid?

We only make files sparse when using multi thread download.

We still preallocate them though so I assume the preallocation is still doing its job on Windows. I presume this is a backwards compatibility thing in Windows - having to enable sparse files - you don't need to do that on Linux and it works just fine with the preallocation.

**Comment by mehrdadn** [comment]
I think they're directly contradictory; I don't see how this could work. The very notion of sparse files (at least on Windows) is to [_not_ allocate disk space for any extent that is not written to](https://docs.microsoft.com/en-us/windows/win32/fileio/sparse-files). So when you write to each extent, the OS has to find some random chunk of space on the disk to allocate to the file... which is exactly the same problem we had to begin with.

I'm not sure why or how it might be different on Linux, but at least on Windows, I would expect that this change has actually defeated the original change. In fact, it might have even made it worse, because now these are marked as sparse files and require extra metadata just to keep track of which extents are non-sparse! This is definitely not something you want to assume has improved without testing.

If the blocking on zero-filling is undesirable, then I would suggest one of the following:

- Buffer files in memory until the disk space is preallocated, then flush everything to disk. This is probably nontrivial.

- Provide a flag to avoid preallocation instead.

Either way, I wouldn't use sparse files, since they're not really meant for this. Rather, they're meant for files with large chunks of zeros in them.

Also keep in mind that fragmentation impacts HDDs far worse than zero-filling delay (it's a severe performance degredation on _every_ read vs. a relatively mild, one-time slowdown on the first write), so I don't think you want to make this the default behavior.

**Comment by ferferga** [comment]
@mehrdadn But it's the fragmentation issue such a big deal? Windows automatically defragments HDDs and TRIM SSDs by default weekly. Fragmentation in SSDs doesn't affect performance. However, this makes the double of writes on them, which is even worse.

Common programs that write large files and are Windows only (so they are solely designed for FATXX and NTFS) that write files directly are 7Zip, WinRAR, Movie Maker, Vegas Pro, many torrent programs... Also, cross-platform programs like Google Chrome, Firefox, MKVToolnix, can also be seen starting with files little in size that start to grow once the progress of the download/write operation goes.

**Comment by mehrdadn** [comment]
@ferferga I'm talking about large downloads to external HDDs (just like you), not tiny downloads to SSDs. These disks aren't connected to the computer 24/7 to let Windows defragment them. 

You can't really argue that a 2x contiguous write on an SSD is worse than the wear that weekly defragmentation would incur on an HDD. Though we weren't talking about SSDs in the first place.

Also note that as I explained in my post, this had resulted in 300+ fragments even for a 100MB file. There's not just a little bit of fragmentation, that is absolutely crazy wear on an HDD.

**Comment by ferferga** [comment]
@mehrdadn Hmm, that's true, that's excessive. Missed that out. Anyway, I don't think we should bring all Windows users into the same boat: if this method is kept, I think we should at least differentiate between SSDs and HDDs.

Also, I think sparse files are a good solution to the problem. I'm not an expert by any means, as I said in the first comment here. However, that's exactly what hypervisor platforms like VMWare does with disks that are created with the non-dynamic disks: They create a (seemingly) huge file first, but if you right click it and go to "Properties", the actual size in disk is way lower, so they can fill it later (but the space is already reserved to that file).

**Comment by mehrdadn** [comment]
Sparse files are *not* "delayed-allocation" files; they're not solutions to this problem at all. They're for sparse data (i.e. with zeros). The documentation is pretty clear on this. Virtual disks would be *actually* sparse until/unless they get filled. Whereas these files *definitely* get filled, and get filled *immediately*.

I'm confused what you're trying to argue against though. Both of the 2 proposals [I suggested](#issuecomment-627942946) should work for you, right?

**Comment by ferferga** [comment]
@mehrdadn Yeah, both are good. 

I'm not "trying to argue against" this, especially because I have no idea of go and many of the underlying features of filesystems. I just propose that we can study how those apps I mentioned before (which have *seemingly* writing patterns similar to rclone) work, because none of them have those problems. There *must be a better way*.

**Comment by mehrdadn** [comment]
I see. Have you used `aria2c`? It's a pretty well-known program that runs into exactly this problem and it handles this with a flag ([`--file-allocation`](https://aria2.github.io/manual/en/html/aria2c.html#cmdoption-file-allocation)) to let the user choose the allocation method.

**Comment by mehrdadn** [comment]
Just don't get too excited at its "falloc" feature because it has security implications and cannot work without administrator permissions. (Though it's still an option we can consider for when it's possible.)

**Comment by ferferga** [comment]
@mehrdadn I'm not too sure aria2 is comparable to rclone because aria2 is capable of resuming stopped/cancelled downloads.

**Comment by mehrdadn** [comment]
Why is that relevant?

Also weren't you comparing to virtual disks which have far, _far_ less in common with downloads..?

**Comment by ferferga** [comment]
@mehrdadn Because they're archival files with a similar nature to 7z, zip and so on that usually are large and programs that create them don't usually do preallocation.

I guess it's needed for torrents, as files are often downloaded in multiparts and not in sequential order. If you want to resume that torrent download later, you need to have a somewhat coherent structure to verify that each part is correct.

The source of the data is irrelevant (specially because rclone can have multiple sources of data: RAM, clouds, local filesystem, etc). The difference we're studying is in how the data is written. Torrenting and P2P doesn't have the same I/O pattern that rclone uses, as, afaik, there is no storage system where it doesn't write sequentally, from start to end, unlike most P2P.

Anyway, imho, this whole discussion and comparison between data formats is getting to nowhere: what we must resolve and asks ourselves is how 7zip, WinRAR, etc accomplish this problem.

**Comment by ncw** [comment]
> I think they're directly contradictory; I don't see how this could work. The very notion of sparse files (at least on Windows) is to [_not_ allocate disk space for any extent that is not written to](https://docs.microsoft.com/en-us/windows/win32/fileio/sparse-files). So when you write to each extent, the OS has to find some random chunk of space on the disk to allocate to the file... which is exactly the same problem we had to begin with.

Hmm... I see what you mean.
 
> I'm not sure why or how it might be different on Linux, but at least on Windows, I would expect that this change has actually defeated the original change. In fact, it might have even made it worse, because now these are marked as sparse files and require extra metadata just to keep track of which extents are non-sparse! This is definitely not something you want to assume has improved without testing.

We only set the sparse flag when we are using --multi-thread-streams. This is only when downloading largish files and you can disable it with --multi-thread-streams 0.

Multiple download streams are a big win, however they caused a massive regression on Windows with Window's desire to fill the entire file before the write pointer with 0s before allowing us to write towards the end of the file.

Making them sparse fixed this and made for happy users!

So the users have control of this with the --multi-thread-streams flag. If they turn it off then they'll get pre-allocation as before and if they turn it on they'll get sparse files which are completely filled by the end of the transfer - hopefully NTFS is capable of optimizing that...

> If the blocking on zero-filling is undesirable, then I would suggest one of the following:
> 
> * Buffer files in memory until the disk space is preallocated, then flush everything to disk. This is probably nontrivial.

Won't that cause the extremely long delays while the OS zeros all the file? Or have I mis-understood. That is the problem the users complained about - having to wait while the OS write 50GB of zeros to the disk

Here is an example issue - there were several here and on the forum https://github.com/rclone/rclone/issues/4192

> * Provide a flag to avoid preallocation instead.

We have that already with `--multi-thread-streams 0`

> Either way, I wouldn't use sparse files, since they're not really meant for this. Rather, they're meant for files with large chunks of zeros in them.
> 
> Also keep in mind that fragmentation impacts HDDs far worse than zero-filling delay (it's a severe performance degredation on _every_ read vs. a relatively mild, one-time slowdown on the first write), so I don't think you want to make this the default behavior.

I had a lot of bad experiences sparse files made with bittorrent on Linux. They would write random blocks throughout the file and cause copying these files to be extremely slow! I didn't want to introduce that into rclone, so by default we only download with 4 streams so effectively we cut the file into 4 allocations which doesn't seem too bad to me.

**Comment by mehrdadn** [comment]
@ferferga The I/O pattern of rclone is just like that of a downloader like `aria2c` and very much unlike that of a VM or 7-zip. It's facing exactly the same problem all multithreaded downloaders face, and they all have options to preallocate. Also, you can't just blindly look at another program and copy what they're doing without understanding their reasoning. The programs you keep mentioning (7-zip, VMs, etc.) have **very** good reasons not to preallocate. It's not because of performance. Rather, it's because they _can't_ predict how much space to allocate (especially archival programs) unless they risk wasting a _lot_ of disk space if their guess is wrong, and that could suddenly risk accidentally making the entire computer run out of disk space. By contrast, rclone knows _exactly_ how much to preallocate. The trade-offs are *very* different in all respects between downloaders and VMs and archival utilities and such. You want to look at the most similar programs here, and they're downloaders such as `aria2c`.

**Comment by mehrdadn** [comment]
@ncw Right but the problem is these aren't mutually exclusive; there's no reason to force people to choose between multithreaded downloads and preallocation. What if the user wants multi-threaded streams and doesn't don't need to care about the hangs? Users don't need to care if the files are small (even 100MB takes only a couple seconds on an HDD) and they also don't need to care about responsiveness if they're leaving the program overnight. Users should be able to choose these independently of each other⁠—and with file downloaders, they can. In fact, all downloaders (Free Download Manager, aria2c, etc.) face this problem. They solve it with a dedicated flag. That's why I suggested the same solution.

Users might be "happy" with sparse files, but that's probably because they don't realize what they're losing. Everyone understands a short freeze: it's instant feedback and your computer grinds to a halt. But you need to be quite technical, perceptive, and willing to diagnose an HDD degradation to realize that your HDD is slow because rclone isn't preallocating and instead using NTFS sparse files. And of course, lower HDD life is anything but instant feedback. Users will probably take years to notice that, and they'll have no idea it has anything to do with rclone.

> Won't that cause the extremely long delays while the OS zeros all the file? Or have I mis-understood. That is the problem the users complained about - having to wait while the OS write 50GB of zeros to the disk

It depends how you implement it and what _exactly_ the problem you're trying to solve is. The complaint I saw in this issue was something a little bit more specific:

> _rclone takes like double the time it takes for my bandwidth to download the file just to preallocate it._

As far as rclone goes, problem here is that the _download_ grinds to a halt while the disk is initializing. That's why I was suggesting buffering in memory. That way the download is still happening at full speed while the disk is being initialized. As long as the disk is faster than the network speed (which is generally the case; HDDs can do 100MB/s while we have networks speeds of like 10MB/s here), it will then be very quick to catch up when you flush everything after preallocation.

There are more problems here too, though. One of them (though I'm not going to blame this as the source of the problem) is that @ferferga is already overloading the HDD:

> I know that, if I'm doing other operations in my HDD, stuff is going to be slow in my drive. However, many of the time I get this annoyance the drive is still capable of getting 10 MB/s, which is exactly what my connection provides.

Modern HDDs are supposed to get like 100MB/s, so he's _already_ doing the equivalent of e.g. 5-10 simultaneous downloads on the HDD. That's a lot of continuous load for an HDD; in fact I'd say it's even a lot of continuous load to put on an SSD. But anyway, I'm not trying to judge the user's workload; rather, I'm saying that we need to dig deeper to see which part of the problem is something that rclone can or should try to solve. (As I mentioned before, other downloaders already face this problem and, by default, they just make the user wait for the initialization.)

In fact, I wouldn't be surprised if (for example) part of the slowness is due to fragmentation, which rclone might end up making worse.

Let's keep reading to try to dig further into the issue:

> This is even worse when you have more than 5 transfers of a couple gigabytes each. Rclone will fire 5 threads writing to the disk, so it will be all random reads, which really kills performance. Today, I took like an hour to download 20 GB, something that shouldn't take more than 20 minutes.

I understand this is 1 instance of rclone running 5 threads. But is each of those threads writing to multiple parts of the same file in _parallel_, or is each thread writing one file sequentially? If each of those threads is performing a sequential download, then preallocation shouldn't be pre-_initializing_ the disk, so it shouldn't be incurring a delay. Initialization only occurs if you try to write in the middle of a file when you haven't written before it yet. If you have 5 threads each downloading (say) 4 parts of a file... isn't that itself the problem? Your connection is only 10 MB/s after all, and your files are 1GB. Wouldn't the fix here be to use 10 sections/file for 1 file instead of 5 sections/file for 4 files?

**However**, this said, I think rclone can be much smarter than this by default (though this doesn't preclude providing a flag like `aria2c`). The reality is that fragmentation is only a big problem when the fragments are small. Even Windows itself doesn't consider chunks larger than 64MB fragments as "fragmentation". So one thing you could do is to preallocate like before, but space your download chunks by a smarter amount: for example, for N simultaneous chunks, you might want the chunk sizes to be `min(file_size/N, 64MB)`. This should scale much better to larger file sizes, and it should avoid fragmentation and initialization overhead without having to actually avoid either of them.

What do you think?

**Comment by ferferga** [comment]
@mehrdadn The point you have me about the expected file size vs the unknown filesize made me understand better the differences.

Last side note I would like to give, so you both have it in mind for your final decision in how to handle this matter. It's already mentioned by @mehrdadn but I wanted to reassure it.

I talked about HDDs because it's my usual workflow: I usually do some night's automation with a pair of downloads/uploads running in a single rclone instance to my HDD while it's also doing stuff, as mentioned in my first comment.
This means that, of course, my HDD can't provide the usual 100 MB/s speed. 

Although @mehrdadn mentioned you can't blame the user for the workflow, it's fair to note that it's not always dependent on the workflow: My HDD is a modern one capable of those high speeds, however, I sometimes use an old 2.5' one which only gets 50 MB/s or some USB 2.0 flash drives with rclone as well, which barely gets around 20 MB/s. With busy HDDs, it's normal that users expect rclone to last a little more. In fact, they can still be faster than downloading directly from the browser. However, this is not the case by any means for these situations. In fact, I can say 100% sure that a queue gets completed faster by direct downloading it rather than with rclone when speeds are below 40 MB/s. My HDD is usually fragmented as you said (it's an SSHD though) but many of the times I tried with slow drives they were completely formatted (as I have them lying around, I use them when I don't have anything else)

Those situations are scarce, but they exist. I'm still finding time to try your beta, but I would say that being consistent across all platforms is the best idea (if I understood everything correctly, now Windows will behave in the same way as in Unix systems) by default, so advanced users can use the flag when needed. It's just a pity that huge downloads finish faster in a Raspberry Pi (with slow storage) rather than my desktop with some light I/O usage.

**Comment by mehrdadn** [comment]
@ferferga I do think consistency is good too, but only after quality is accounted for. You shouldn't enforce consistently _bad_ behavior across platforms just for the sake of consistency. After all, the very reason we have multiple platforms is that they all have their own strengths and weaknesses; if you enforce consistent behavior then you miss their strengths and end up with their weaknesses.

I think one thing that is helpful here is to insist on the _problem_ rather than insisti on the _solution_. Insisting on the solution ends up solving that one problem at the expense of other ones. That's why I've offered 4 solutions so far: 1 being my original suggestion, with 3 follow-ups to try to solve the issues you're encountering as well. Ultimately we want to solve problems (both yours, mine, and others'), not to force solutions. That requires understanding the trade-offs before deciding which solution is best.

Your point about flash drives is good. It doesn't imply we need to use sparse files or something like that, though. It just implies we need to consider its trade-offs when coming up with a solution. It's actually interesting because random I/O can be _incredibly_ slow on flash drives, so I wouldn't be surprised if a fast network and a slow flash drive actually end up _slower_ with sparse files. (My guess is this is less common but possible.) If you want to handle flash drives, we can think of solutions, but at first we need to establish that they're at least as important as SSDs and HDDs; otherwise, they shouldn't override our concerns on those.

And one thing worth keeping in mind here is that this isn't just a platform issue but also a filesystem issue. You probably don't want a solution that depends on sparse file support to be usable unless there isn't a better option. Writing to (say) exFAT shouldn't suddenly lose performance because of lack of sparse file support. Yet unlike the sparse file solution, both the buffering + chunking solutions, and also the flag solution that lets you choose the allocation method independently from the concurrency, should solve these problems (including yours) pretty consistently on every file system and platform.

(P.S.: My experience with SSHDs hasn't been great. Your drive might die young, and recovering data from it might be much harder than from a plain HDD. Just wanted to warn you in case you weren't aware!)

**Comment by ncw** [comment]
@mehrdadn wrote:
> As far as rclone goes, problem here is that the _download_ grinds to a halt while the disk is initializing. That's why I was suggesting buffering in memory. That way the download is still happening at full speed while the disk is being initialized

Rclone does this already. Set the buffer size with `--buffer-size` - the default is 16M.

> **However**, this said, I think rclone can be much smarter than this by default (though this doesn't preclude providing a flag like `aria2c`).

A flag is a good idea. What does `aria2c` call it? (I like consistency!). It would be to disable setting the sparse bit on Windows (there is no such thing on Unix platforms).

> The reality is that fragmentation is only a big problem when the fragments are small. Even Windows itself doesn't consider chunks larger than 64MB fragments as "fragmentation". So one thing you could do is to preallocate like before, but space your download chunks by a smarter amount: for example, for N simultaneous chunks, you might want the chunk sizes to be `min(file_size/N, 64MB)`. This should scale much better to larger file sizes, and it should avoid fragmentation and initialization overhead without having to actually avoid either of them.

The rclone fragments are large...  By default

      --multi-thread-cutoff SizeSuffix   Use multi-thread downloads for files above this size. (default 250M)
      --multi-thread-streams int         Max number of streams to use for multi-thread downloads. (default 4)

So that means the smallest file we will attempt multithread download for is 250M. Rclone basically allocates [1 stream per 250M of file up to 4 streams](https://rclone.org/docs/#multi-thread-streams-n). We space the writers evenly through the file so the minimum chunk size will be 125M for a 250M file by default.

**Comment by mehrdadn** [comment]
> @mehrdadn wrote:
> 
> > As far as rclone goes, problem here is that the _download_ grinds to a halt while the disk is initializing. That's why I was suggesting buffering in memory. That way the download is still happening at full speed while the disk is being initialized
> 
> Rclone does this already. Set the buffer size with `--buffer-size` - the default is 16M.

Right, so I guess 16M is nowhere near enough. The size probably needs to scale somehow with RAM, the file size, and/or any buffering space available on disk (if any), unless we use a solution that doesn't depend on this.

> > **However**, this said, I think rclone can be much smarter than this by default (though this doesn't preclude providing a flag like `aria2c`).
> 
> A flag is a good idea. What does `aria2c` call it? (I like consistency!). It would be to disable setting the sparse bit on Windows (there is no such thing on Unix platforms).

It has a flag for the file allocation method, called [`--file-allocation`](https://aria2.github.io/manual/en/html/aria2c.html#cmdoption-file-allocation). I'm not sure which options (if any) use sparse files, but I haven't been under the impression that any of them do. However, the _fastest_ option is `falloc`, which I strongly recommend allocating. It's nearly instantaneous. The only downsides is it might not work without high privileges, due to the fact that it makes the system leak disk contents to the program. But when security isn't an issue (it isn't for home users; it's only an issue for servers where the users aren't admins and can't trust each other) it is the option every user would want, and it doesn't require particularly sophisticated FS support.

> The rclone fragments are large... By default

I don't think we're talking about the same fragments, or this wouldn't square with the fact that I had seen a 100MB file get broken into 300 pieces. This isn't about the fragments you request from the *server* (which go from the network to your RAM), but the fragments you write to the client's *disk* (which go from your RAM to the disk). I think you want them to be nice and large chunks every time if the disk space isn't already preallocated (which it wouldn't be with the scenario I was describing int his section).

> We space the writers evenly through the file so the minimum chunk size will be 125M for a 250M file by default.

Yeah so I was saying distributing them evenly is intuitive but doesn't seem like a great idea. That's what's causing the huge delay. For a 50GB you just start a thread writing at offset=25GB and then that triggers initialization, but there's no need to do that. In this example you want to preallocate the file (so fragmentation is minimized) and then maybe have threads start writing at offsets 0MB, 250MB, 500MB, 750MB, etc... that way they don't have to wait for 25GB of initialization.

**Comment by ncw** [comment]
> It has a flag for the file allocation method, called [`--file-allocation`](https://aria2.github.io/manual/en/html/aria2c.html#cmdoption-file-allocation).

That is a useful reference thank you :-)

> I'm not sure which options (if any) use sparse files, but I haven't been under the impression that any of them do. However, the _fastest_ option is `falloc`, which I strongly recommend implementing in rclone.

We've already implement `falloc` so I think this problem is solved for non Windows platforms.

Unix

https://github.com/rclone/rclone/blob/4006345cfbe8ef81554b0fc79aaeb765e9fbc064/lib/file/preallocate_unix.go#L22

Windows

https://github.com/rclone/rclone/blob/4006345cfbe8ef81554b0fc79aaeb765e9fbc064/lib/file/preallocate_windows.go#L36

And here is where they both get called

https://github.com/rclone/rclone/blob/4006345cfbe8ef81554b0fc79aaeb765e9fbc064/backend/local/local.go#L1121-L1129

For windows platforms we need to set the Sparse flag also otherwise we get that delay when then OS writes everything to 0 up to the file pointer. I don't know why that is - maybe the prealloc isn't working properly?

https://github.com/rclone/rclone/blob/4006345cfbe8ef81554b0fc79aaeb765e9fbc064/lib/file/preallocate_windows.go#L86

> It's nearly instantaneous and it'd work fine for most users. (It corresponds to setting `FileValidDataLengthInformation` in Windows, which just expands the file size while taking whatever data is there in the underlying disk.) The only downside is it might not work without high privileges, due to the fact that it makes the system leak disk contents to the program. But when security isn't an issue (it isn't for home users; it's only an issue for servers where the users aren't admins and can't trust each other) it is the option every user would want, and it doesn't require particularly sophisticated FS support.

Fast but leaky! Our falloc on Windows doesn't work like that though...

> I don't think we're talking about the same fragments, or this wouldn't square with the fact that I had seen a 100MB file get broken into 300 pieces.

Ah... How do you measure the fragmentation for a file?

> This isn't about the fragments you request from the _server_ (which go from the network to your RAM), but the fragments you write to the client's _disk_ (which go from your RAM to the disk). I think you want them to be nice and large chunks every time if the disk space isn't already preallocated (which it wouldn't be with the scenario I was describing in this section).

I was assuming that if you are writing at 4 points in the file then you'll be making 4 fragments, but it seems I was wrong about that!

> Yeah so I was saying distributing them evenly is intuitive but doesn't seem like a great idea. That's what's causing the huge delay. For a 50GB you just start a thread writing at offset=25GB and then that triggers initialization, but there's no need to do that. In this example you want to preallocate the file (so fragmentation is minimized) and then maybe have threads start writing at offsets 0MB, 250MB, 500MB, 750MB, etc... that way they don't have to wait for 25GB of initialization.

That is a good idea... There are tradeoffs here though. Rclone would like to be reading large streams of data from the server - that is the most efficient thing to do. Each TCP connection takes a while to establish and get going. However if the chunk size was ~250MB I'm sure that is large enough. That would make a lot more requests for a large file which some providers (s3) would charge you for and some (Google Drive) might ban you for.

**Comment by mehrdadn** [comment]
> I was assuming that if you are writing at 4 points in the file then you'll be making 4 fragments, but it seems I was wrong about that!

Yeah, not on Windows, unless you're dealing with sparse files. For normal files, Windows just "extends" the initialized portion of the file to that point, which can take forever if you're writing at offset 25GB and none of the file has been initialized (regardless of whether it has been allocated).

I thought I'd share a couple of experiments to illustrate what happens in reality with sparse files. Here's what happens on Windows. Note that you can see the fragmentation via `fsutil file queryExtents`, but there's also [`contig -a`](https://docs.microsoft.com/en-us/sysinternals/downloads/contig) which is similar.

```
C:\>fsutil file createNew temp 0 && fsutil sparse setFlag temp 1 && fsutil file setEOF temp 134217728 && (echo.>>temp) && fsutil file setEOF temp 268435456 && (echo.>>temp) && fsutil file queryExtents temp && del temp
File C:\temp is created
File C:\temp eof set
File C:\temp eof set
VCN: 0x0        Clusters: 0x8000     LCN: 0xffffffffffffffff
VCN: 0x8000     Clusters: 0x10       LCN: 0xa4b91e
VCN: 0x8010     Clusters: 0x7ff0     LCN: 0xffffffffffffffff
VCN: 0x10000    Clusters: 0x10       LCN: 0xa4cb52
```

LCN is the logical cluster number (i.e. the block number relative to the beginning of the volume), and VCN is the LBA aka the virtual cluster number (the block number relative to the beginning of the file).

Notice that there are 0x7ff0 clusters between the two allocated extents virtually, but only 0xa4cb52 - (0xa4b91e + 0x10) = 0x1224 clusters allocated to the file on the volume. This means the file will never end up contiguous. Extents can even end up out-of-order as a result.

Here's what the equivalent would look like on Linux:

```
$ fallocate -l 134217728 temp && fallocate -p -l 134217728 temp && (echo "">>temp) && fallocate -o 134217729 -l 134217728 temp && fallocate -p -o 134217729 -l 134217728 temp && (echo "">>temp) && sudo hdparm --fibmap temp && rm -f temp

temp:
 filesystem blocksize 4096, begins at LBA 2048; assuming 512 byte sectors.
 byte_offset  begin_LBA    end_LBA    sectors
   134217728  195430400  195430407          8
   268435456  196298752  196298759          8
```

It's a similar story here. There are (196298752 - 195430407) * 4096 = 3556741120 bytes between the FS blocks, but only 268435456 - (134217728 + 1) = 134217727 in the file. Again, the file cannot be contiguous.

So we conclude sparse files have fragmentation problems.

Now, the nice thing about `fallocate` (at least on ext4) is that it seems to support blocks that are *allocated yet uninitialized*. And the system will return zeros if you try to read such blocks.

NTFS doesn't quite support this. What it *does* support is something called Windows calls a "valid data length", which is the length of the file (starting from offset 0) which is assumed to contain initialized data on the disk. On Windows, the "fallocate" method of aria2c *directly sets the valid data length* ([`SetFileValidData`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-setfilevaliddata)), which is like setting the file length while taking whatever is on the disk as the file contents.

What this means for you is that you have the following options:

- To minimize fragmentation *and* initialization time, **you want to call `SetFileValidData`**, and I think this needs to be _after_ setting `FileAllocationInformation` rather than before (but I'm not sure). However, this requires `SeManageVolumePrivilege`, since it can leak underlying disk contents. I think you need to be an administrator and **also** request this privilege explicitly. The easiest way is to call [`RtlAdjustPrivilege(28 /*SeManageVolumePrivilege*/, TRUE, FALSE, &wasEnabled)`](https://source.winehq.org/WineAPI/RtlAdjustPrivilege.html) from `ntdll.dll` at program startup time. If the call succeeds (returns zero), then you know you can call `SetFileValidData` and don't need to do anything else. If it fails, and the user has passed `--file-allocation=falloc`, then I suggest aborting: they've probably forgotten to use "Run As Administrator", and you want to let them know. If they haven't specified anything, however, then you probably want to be "smart" and try the other options below.

- You can just keep going and make the user wait for initialization during download. It's what download programs typically do. It avoids fragmentation as much as possible and it works on all file systems. But the waiting time can be prohibitive for a huge file. If you do this, you can also consider temporarily [lowering the I/O priority](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/ns-wdm-_file_io_priority_hint_information) to `IoPriorityLow`, and restoring it to `IoPriorityNormal` before actually downloading the file. Or you can provide a flag to the user to do this manually if they're interested. This will ensure other activity doesn't grind to a halt while the file is getting initialized.

- You can use the buffering technique I mentioned earlier (with, say, 64MB or 256MB or even 1GB chunks). This allows for a bit of fragmentation, but with very large fragments, so it's unlikely to be a problem for anybody. I think this is the smartest solution if `SetFileValidData` fails, but I can't speak to the charge/ban issue. You'll need to figure out (possibly on a case-by-case basis) if this solution makes sense for the given remote. I suspect if your chunks are large (say, ≥ 256MB?) it shouldn't be a problem, but I don't really know.

Hope this clarified everything you were asking about!

**Comment by ncw** [comment]
> Hope this clarified everything you were asking about!

Thank you! That is very helpful :-)

I'll make a new issue about this and take a look in a little while (trying to get the current release out at the moment!).

For the current release I could do one or more of

1. caveat the docs on multi thread download about sparse files on Windows
2. allow the sparse file setting to be disabled on Windows
3. default to not using multihread download on Windows

I think I should definitely do 1. Option 2. is pretty easy. Option 3 might be controversial.

Thoughts?

**Comment by mehrdadn** [comment]
Sounds good, yeah I agree, options 1-2 should be good. For option 3, I'd keep the behavior the same as the last release; it's probably best not to cause a silent change for now. If it's easy to implement, you could also consider printing a warning on Windows (and/or a note in the help text) if the user hasn't explicitly specified any flags for multithreaded downloads or sparse files (and hasn't silenced via `--quiet`) to let them know sparse files might cause fragmentation? Up to you haha.
